File: backend/src/main/java/br/ufpr/sept/so2/modules/presenca/infrastructure/persistence/PresencaJpaAdapter.java
package br.ufpr.sept.so2.modules.presenca.infrastructure.persistence;

import br.ufpr.sept.so2.modules.presenca.application.ports.PresencaRepository;
import br.ufpr.sept.so2.modules.presenca.domain.FasePresenca;
import br.ufpr.sept.so2.modules.presenca.domain.Presenca;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.UUID;

@Component
public class PresencaJpaAdapter implements PresencaRepository {

    private final PresencaJpaRepository jpaRepository;

    public PresencaJpaAdapter(PresencaJpaRepository jpaRepository) {
        this.jpaRepository = jpaRepository;
    }

    @Override
    public Presenca save(Presenca presenca) {
        return jpaRepository.save(PresencaJpaEntity.fromDomain(presenca)).toDomain();
    }

    @Override
    public boolean existsByEventoUsuarioFase(UUID eventoId, UUID usuarioId, FasePresenca fase) {
        return jpaRepository.existsByEventoIdAndUsuarioIdAndFase(eventoId, usuarioId, fase.name());
    }

    @Override
    public boolean existsByEventoAndDeviceDeOutroUsuario(UUID eventoId, String deviceUuid, UUID usuarioId) {
        return jpaRepository.existsByEventoIdAndDeviceUuidAndUsuarioIdNot(eventoId, deviceUuid, usuarioId);
    }

    @Override
    public List<Presenca> findByEventoAndUsuario(UUID eventoId, UUID usuarioId) {
        return jpaRepository.findByEventoIdAndUsuarioId(eventoId, usuarioId).stream()
                .map(PresencaJpaEntity::toDomain)
                .toList();
    }

    @Override
    public long countByEvento(UUID eventoId) {
        return jpaRepository.countByEventoId(eventoId);
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/presenca/infrastructure/persistence/PresencaJpaAdapter.kt
package br.ufpr.sept.so2.modules.presenca.infrastructure.persistence

import br.ufpr.sept.so2.modules.presenca.application.ports.PresencaRepository
import br.ufpr.sept.so2.modules.presenca.domain.FasePresenca
import br.ufpr.sept.so2.modules.presenca.domain.Presenca
import org.springframework.stereotype.Component
import java.util.UUID

@Component
class PresencaJpaAdapter(
    private val jpaRepository: PresencaJpaRepository,
) : PresencaRepository {

    override fun save(presenca: Presenca): Presenca = jpaRepository.save(PresencaJpaEntity.fromDomain(presenca)).toDomain()

    override fun existsByEventoUsuarioFase(eventoId: UUID, usuarioId: UUID, fase: FasePresenca): Boolean =
        jpaRepository.existsByEventoIdAndUsuarioIdAndFase(eventoId, usuarioId, fase.name)

    override fun existsByEventoAndDeviceDeOutroUsuario(eventoId: UUID, deviceUuid: String, usuarioId: UUID): Boolean =
        jpaRepository.existsByEventoIdAndDeviceUuidAndUsuarioIdNot(eventoId, deviceUuid, usuarioId)

    override fun findByEventoAndUsuario(eventoId: UUID, usuarioId: UUID): List<Presenca> =
        jpaRepository.findByEventoIdAndUsuarioId(eventoId, usuarioId).map { it.toDomain() }

    override fun countByEvento(eventoId: UUID): Long = jpaRepository.countByEventoId(eventoId)
}
